workflow/controller/controller.go

```diff
@@ -1270,71 +1238,150 @@ func (wfc *WorkflowController) newPodInformer(ctx context.Context) (cache.Shared
-					key, _ := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
-					namespace, name, _ := cache.SplitMetaNamespaceKey(key)
-					delete(wfc.executorPlugins[namespace], name)
-					log.WithField("namespace", namespace).WithField("name", name).Info("Executor plugin removed")
-				},
+	//nolint:errcheck // the error only happens if the informer was stopped, and it hasn't even started (https://github.com/kubernetes/client-go/blob/46588f2726fa3e25b1704d6418190f424f95a990/tools/cache/shared_informer.go#L580)
+	indexInformer.AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			cmMeta, err := meta.Accessor(obj)
+			if err != nil {
+				log.WithError(err).
+					Error("failed to get configmap metadata")
+				return false
+			}
+
+			return isPluginCM(cmMeta)
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				cm := obj.(*apiv1.ConfigMap)
+				wfc.applyPluginCM(cm, "added")
 			},
-		})
-		if err != nil {
-			return nil, err
+			UpdateFunc: func(_, obj interface{}) {
+				cm := obj.(*apiv1.ConfigMap)
+				wfc.applyPluginCM(cm, "updated")
+			},
+			DeleteFunc: func(obj interface{}) {
+				cm := obj.(*apiv1.ConfigMap)
+				wfc.deletePluginCM(cm)
+			},
+		},
+	})
+	return indexInformer
+}
+
+func isPluginCM(cmMeta metav1.Object) bool {
+	return cmMeta.GetLabels()[common.LabelKeyConfigMapType] == common.LabelValueTypeConfigMapExecutorPlugin
+}
+
+func (wfc *WorkflowController) applyPluginCM(cm *apiv1.ConfigMap, verb string) {
+	p, err := plugin.FromConfigMap(cm)
+	if err != nil {
+		log.WithField("namespace", cm.GetNamespace()).
+			WithField("name", cm.GetName()).
+			WithError(err).
+			Error("failed to convert configmap to plugin")
+		return
+	}
+	if _, ok := wfc.executorPlugins[cm.GetNamespace()]; !ok {
+		wfc.executorPlugins[cm.GetNamespace()] = map[string]*spec.Plugin{}
+	}
+	wfc.executorPlugins[cm.GetNamespace()][cm.GetName()] = p
+	log.WithField("namespace", cm.GetNamespace()).
+		WithField("name", cm.GetName()).
+		Infof("Executor plugin %s", verb)
+}
+
+func (wfc *WorkflowController) deletePluginCM(cm *apiv1.ConfigMap) {
+	key, _ := cache.DeletionHandlingMetaNamespaceKeyFunc(cm)
+	namespace, name, _ := cache.SplitMetaNamespaceKey(key)
+	delete(wfc.executorPlugins[namespace], name)
+	log.WithField("namespace", namespace).WithField("name", name).Info("Executor plugin removed")
+}
+
+var watchControllerSemaphoreConfigMaps = os.Getenv("WATCH_CONTROLLER_SEMAPHORE_CONFIGMAPS") != "false"
+
+func (wfc *WorkflowController) newConfigMapControllerInformer(ctx context.Context) cache.SharedIndexInformer {
+	indexInformer := v1.NewFilteredConfigMapInformer(wfc.kubeclientset, wfc.GetNamespace(), 20*time.Minute, nil, func(opts *metav1.ListOptions) {
+		opts.FieldSelector = fields.OneTermEqualSelector(metav1.ObjectNameField, wfc.configController.GetName()).String() // only the controller configmap
+	})
+
+	if !watchControllerSemaphoreConfigMaps {
```

Comment:
I'm confused about this. Is it supposed to be here, or was it copy/pasted from below?

Reply:
It is supposed to be there. If the flag is set, no event handlers are added to semaphore or controller informers.

They're together only because the flag was implemented prior to them being split (it was implemented as a workaround for the main bug this PR aims to fix)

Reply:
Oh jeez. Backward compatibility can be so annoying sometimes. :) I suppose some people do intend to not watch their Controller ConfigMap and are using that environment variable? In that case, can we make it more clear in the code through comments and/or naming that it represents both things?

Reply:
> I suppose some people do intend to not watch their Controller ConfigMap and are using that environment variable?

Yes, https://github.com/argoproj/argo-helm/issues/2622 is one example, although I'm not committed to supporting that. I wanted to leave this flag in just-in-case there are remaining issues. The hot reload has also been buggy before, e.g. #11177, #11542, #11541, so it is potentially useful as a workaround if that reoccurs.

> In that case, can we make it more clear in the code through comments and possibly naming that it represents both things?

Sure I can add a comment. I made sure it was described in the docs in #12622 and #12400, but code comments are missing.

The naming follows the env variable exactly as the codebase conventionally does, so I wouldn't change that though.

Reply:
okay, sounds good

Reply:
[added](https://github.com/argoproj/argo-workflows/pull/11855/commits/05ee735a985473e4f15b1fc642f6ff96571aa7c5) comment

Also fixed merge conflicts (it was with my own PR that stemmed from a finding in this PR: #12917. just had to choose this PR's version for each conflict)